### ISSUE TITLE
Deserializing an Undefined Bool value results in False, not Bool.

### ISSUE THREAD

**Issue by Xliff** [jonathanstowe/XML-Class#4]
If I have the following invocation:

```
perl6 -Ilib -e 'use XML::Class; class A does XML::Class[xml-element => "A"] { has Bool $.b is xml-element is rw; }; my $a = A.new; $a.b = Nil; $a.gist.say; my $xml = $a.to-xml; $xml.say; my $b = A.from-xml($xml); $b.gist.say'
```

I get the following output:

```
A.new(b => Bool)
<?xml version="1.0"?><A><b>false</b></A>
A.new(b => Bool::False)
```

I was expecting that the XML would output `<b />`. If I try removing the blank element via the `xml-skip-null` trait, the XML looks right:

```
<?xml version="1.0"?><A/>
```

However, if I deserialize, I still get Bool::False, rather than Bool.

```
A.new(b => Bool::False)
```

Is this the intended behavior, or can XML::Class be adjusted to return a Bool, instead of Bool::False?

**Comment by jonathanstowe** [comment]
Well, it's deliberate, though I can see your point.  

You can of course over-ride the serialisation/deserialisation behaviour with ```xml-serialise``` and ```xml-deserialise```

**Comment by Xliff** [comment]
I've thought about that, but the documentation is a bit lax, since it only focuses on strings. How would that work for Bool, or better yet... arbitrary objects?

Thanks.

**Comment by jonathanstowe** [comment]
I'm not quite following you.

Anyhow this does what you appear to want:

```
    use XML::Class;
    class Booly does XML::Class {
        sub bool-out(Bool $v --> Str ) {
            $v.defined ?? $v ?? 'true' !! 'false' !! Str;
        }
        sub bool-in(Str $v --> Bool ) {
            $v.defined ?? $v eq 'true' ?? True !! False !! Bool
        }

        has Bool $.bool-element is xml-element is xml-serialise(&bool-out) is xml-deserialise(&bool-in);
    }

    my $xml = Booly.new.to-xml;
    say $xml;
    say Booly.from-xml($xml);
```

**Comment by Xliff** [comment]
Excellent! Thanks. I will check this and get back to you. If it works, I will close the issue.

**Comment by Xliff** [comment]
@jonathanstowe - OK, that covers the Bool case, but what about for an object?

Given the following class:

```
class TestClass is export
  does XML::Class[xml-element => 'item']
{
  has Str    $.attr1    is xml-element          is rw; 
  has Str    $.attr2    is xml-element          is rw; 
}
```

...what would be the best way to write a custom de/serializer

**Comment by jonathanstowe** [comment]
I'm not quite sure what you mean.  The idea behind the (de)serialisation hooks is to convert some type (for example a Date, a Version or some such,) that can meaningfully be expressed as "simple content" (basically an XML value,)  In the way that it is designed a Class implies an XML "complex type"  which will be a set of nested elements with attributes and so forth, the aim is to be able to be able to deal with anything that can be described by an XML schema.  So a (de)serialisation hook doesn't really make sense for a custom type unless you are proposing to pack/unpack it to/from a string.

If, on the other hand, you have some custom type that needs to represent some specific XML and you have a free hand as to the interface of the class then you should be able to achieve that with the existing facilities.  If both Class interface and XML are not in your control then it might be better to create a class to deal with XML with a factory method that creates the second class that conforms to your interface based on the data from the XML.

If all else fails you can provide your own candidates for ```to-xml``` and ```from-xml``` that return and accept an ```XML::Element``` respectively,  but I'd probably advise against that as you'd have to everything yourself.

**Comment by Xliff** [comment]
This has mostly been addressed by #3. Closing issue.
